### PR TITLE
Fetch Cargo manually from new archives

### PR DESCRIPTION
**Pull request by alexcrichton**

### CODE REVIEW

**Comment by alexcrichton** [comment]
The tl;dr; of the scheme I'm envisioning is:

* Cargo has a bunch of `rust-1.X.Y` branches
* Rust version `1.X.Y` is paired with the Cargo produced from the branch `rust-1.X.Y`
* When you r+ a PR on Cargo, it uploads all artifacts that were built and tested
* As a result, all merge commits on the `rust-1.X.Y` are guaranteed to have artifacts uploaded
* Backporting is therefore just a PR against `rust-1.X.Y` with a `@bors: r+`
* On this, the packaging side, we just pick up the latest commit on `rust-1.X.Y` with an API call to GitHub, and then we download the right artifact from there.

I haven't *quite* figure out how to deal with Cargo's versioning just yet. Cargo builds all release artifacts on the `auto-cargo` branch and doesn't have a great way of distinguishing what channel the build is for. Similarly, if we have different names for artifacts in the cargo upload directories we'd need to discover that here as well.

What I envision is that all `rust-1.X.Y` branches in Cargo are built with `--release-channel=stable` for Cargo. The artifacts are then named with `cargo-<cargo version>-<target>.tar.gz`. I guess we can upload some file in each folder with a deterministic name with the contents being `<cargo version>`?

@brson, thoughts on this?

File: fetch-inputs.py [hunk 1]
```diff
@@ -33,8 +33,9 @@
     sys.exit(1)
 
 SERVER_ADDRESS = os.getenv("RUST_DIST_SERVER", "https://static.rust-lang.org")
+CARGO_SERVER_ADDRESS = os.getenv("CARGO_DIST_SERVER",
+        "https://s3-us-west-1.amazonaws.com/rust-lang-cargo-dev/cargo-master")
```
Comment (brson) [inline fetch-inputs.py]: What does the naming of this bucked signify? It looks like a "dev" bucket for "master", but doesn't this script fetch from a production bucket for other branches than master?

**Comment by brson** [comment]
This all sounds fine to me. r+
